File: src/main/java/me/cjcrafter/armormechanics/Command.java
package me.cjcrafter.armormechanics;

import me.deecaad.core.MechanicsCore;
import me.deecaad.core.commands.*;
import me.deecaad.core.commands.arguments.EntityListArgumentType;
import me.deecaad.core.commands.arguments.MapArgumentType;
import me.deecaad.core.commands.arguments.StringArgumentType;
import me.deecaad.core.compatibility.CompatibilityAPI;
import me.deecaad.core.utils.EnumUtil;
import me.deecaad.core.utils.StringUtil;
import me.deecaad.weaponmechanics.WeaponMechanics;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.inventory.EntityEquipment;
import org.bukkit.inventory.EquipmentSlot;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.PluginDescriptionFile;

import java.util.*;
import java.util.function.Function;

import static org.bukkit.ChatColor.*;

@SuppressWarnings("unchecked")
public class Command {

    public static final char SYM = '\u27A2';

    public static Function<CommandData, Tooltip[]> ARMOR_SUGGESTIONS = (data) -> {
        return ArmorMechanics.INSTANCE.armors.keySet().stream().map(Tooltip::of).toArray(Tooltip[]::new);
    };

    public static Function<CommandData, Tooltip[]> SET_SUGGESTIONS = (data) -> {
        return ArmorMechanics.INSTANCE.sets.keySet().stream().map(Tooltip::of).toArray(Tooltip[]::new);
    };

    public static void register() {

        MapArgumentType giveDataMap = new MapArgumentType()
                .with("dontEquip", MapArgumentType.INT(0, 1))
                .with("forceEquip", MapArgumentType.INT(0, 1))
                .with("preventRemove", MapArgumentType.INT(0, 1));

        CommandBuilder command = new CommandBuilder("am")
                .withAliases("armor", "armormechanics")
                .withPermission("armormechanics.admin")
                .withDescription("ArmorMechanics' main command")

                .withSubcommand(new CommandBuilder("give")
                        .withPermission("armormechanics.commands.give")
                        .withDescription("Gives the target(s) the requested armor")
                        .withArgument(new Argument<>("targets", new EntityListArgumentType()).withDesc("Which target(s) to give the armor"))
                        .withArgument(new Argument<>("armor", new StringArgumentType()).withDesc("Which armor to give").replace(ARMOR_SUGGESTIONS))
                        .withArgument(new Argument<>("data", giveDataMap, new HashMap<>()).withDesc("How to equip the armor"))
                        .executes(CommandExecutor.any((sender, args) -> {
                            give(sender, (List<Entity>) args[0], (String) args[1], (HashMap<String, Object>) args[2]);
                        })))

                .withSubcommand(new CommandBuilder("get")
                        .withPermission("armormechanics.commands.get")
                        .withDescription("Gives you the requested armor")
                        .withArgument(new Argument<>("armor", new StringArgumentType()).withDesc("Which armor to give").replace(ARMOR_SUGGESTIONS))
                        .executes(CommandExecutor.entity((sender, args) -> {
                            give(sender, Collections.singletonList(sender), (String) args[0], new HashMap<>());
                        })))

                .withSubcommand(new CommandBuilder("giveset")
                        .withPermission("armormechanics.commands.giveset")
                        .withDescription("Gives the target(s) the requested set of armor")
                        .withArgument(new Argument<>("targets", new EntityListArgumentType()).withDesc("Which target(s) to give the set"))
                        .withArgument(new Argument<>("set", new StringArgumentType()).withDesc("Which set of armor to give").replace(SET_SUGGESTIONS))
                        .withArgument(new Argument<>("data", giveDataMap, new HashMap<>()).withDesc("How to equip the armor"))
                        .executes(CommandExecutor.any((sender, args) -> {
                            ArmorSet set = ArmorMechanics.INSTANCE.sets.get((String) args[1]);
                            if (set.getHelmet() != null) give(sender, (List<Entity>) args[0], set.getHelmet(), (Map<String, Object>) args[2]);
                            if (set.getChestplate() != null) give(sender, (List<Entity>) args[0], set.getChestplate(), (Map<String, Object>) args[2]);
                            if (set.getLeggings() != null) give(sender, (List<Entity>) args[0], set.getLeggings(), (Map<String, Object>) args[2]);
                            if (set.getBoots() != null) give(sender, (List<Entity>) args[0], set.getBoots(), (Map<String, Object>) args[2]);
                        })))

                .withSubcommand(new CommandBuilder("clear")
                        .withPermission("armormechanics.commands.clear")
                        .withDescription("Clears the target's armor")
                        .withArgument(new Argument<>("targets", new EntityListArgumentType()).withDesc("Which target(s) to clear"))
                        .withArgument(new Argument<>("slots", new StringArgumentType().withLiteral("*"), "*").withDesc("Which slot(s) to clear").replace(SuggestionsBuilder.from("head", "chest", "legs", "feet")))
                        .executes(CommandExecutor.any((sender, args) -> {
                            clear(sender, (List<Entity>) args[0], (String) args[1]);
                        })))

                .withSubcommand(new CommandBuilder("reload")
                        .withPermission("armormechanics.commands.reload")
                        .withDescription("Reloads the plugin's configurations")
                        .executes(CommandExecutor.any((sender, args) -> {
                            ArmorMechanics.INSTANCE.reload();
                            sender.sendMessage(GREEN + "Reloaded ArmorMechanics");
                        })))

                .withSubcommand(new CommandBuilder("info")
                        .withPermission("armormechanics.commands.info")
                        .withDescription("Displays information about ArmorMechanics")
                        .executes(CommandExecutor.any((sender, args) -> {
                            info(sender);
                        })));

        command.registerHelp(HelpCommandBuilder.HelpColor.from(GOLD, GRAY, SYM));
        command.register();
    }

    public static void give(CommandSender sender, List<Entity> entities, String title, Map<String, Object> data) {

        // Since we want to ignore spelling/capitalization errors, we should
        // make sure the given 'title' matches to an actual armor-title.
        List<String> startsWith = new ArrayList<>();
        Set<String> options = ArmorMechanics.INSTANCE.armors.keySet();
        for (String temp : options) {
            if (temp.toLowerCase(Locale.ROOT).startsWith(title.toLowerCase(Locale.ROOT)))
                startsWith.add(title);
        }

        title = StringUtil.didYouMean(title, startsWith.isEmpty() ? options : startsWith);
        ItemStack armor = ArmorMechanics.INSTANCE.armors.get(title);
        EquipmentSlot slot = ArmorMechanicsAPI.getEquipmentSlot(armor.getType());

        boolean dontEquip = 1 == (int) data.getOrDefault("dontEquip", 0);
        boolean force = 1 == (int) data.getOrDefault("forceEquip", 0);
        boolean preventRemove = 1 == (int) data.getOrDefault("preventRemove", 0);

        if (!force && preventRemove) {
            sender.sendMessage(RED + "When using preventRemove, forceEquip must also be enabled!");
        }

        for (Entity entity : entities) {
            if (!entity.getType().isAlive())
                continue;

            LivingEntity living = (LivingEntity) entity;
            EntityEquipment equipment = living.getEquipment();

            if (!dontEquip && force) {
                if (preventRemove) {
                    ItemStack clone = armor.clone();
                    CompatibilityAPI.getNBTCompatibility().setInt(clone, "ArmorMechanics", "prevent-remove", 1);
                    ArmorMechanicsAPI.setItem(equipment, slot, clone);
                    return;
                }

                ArmorMechanicsAPI.setItem(equipment, slot, armor.clone());
                return;
            }

            if (!dontEquip && ArmorMechanicsAPI.getItem(equipment, slot) == null) {
                ArmorMechanicsAPI.setItem(equipment, slot, armor.clone());
                return;
            }

            if (living.getType() == EntityType.PLAYER) {
                Player player = (Player) living;
                HashMap<Integer, ItemStack> overflow = player.getInventory().addItem(armor.clone());

                if (!overflow.isEmpty()) {
                    sender.sendMessage(ChatColor.RED + player.getName() + "'s inventory was full!");
                }
            }
        }
    }

    public static void clear(CommandSender sender, List<Entity> targets, String slots) {
        EquipmentSlot slot = "*".equals(slots) ? null : EquipmentSlot.valueOf(StringUtil.didYouMean(slots, EnumUtil.getOptions(EquipmentSlot.class)).toUpperCase(Locale.ROOT));

        sender.sendMessage(GREEN + "Removing armor from " + targets.size() + (targets.size() == 1 ? "entity" : "entities"));

        for (Entity target : targets) {
            if (!target.getType().isAlive())
                continue;

            LivingEntity entity = (LivingEntity) target;
            EntityEquipment equipment = entity.getEquipment();

            if (slot == null) {
                equipment.setBoots(null);
                equipment.setLeggings(null);
                equipment.setChestplate(null);
                equipment.setHelmet(null);
            } else {
                ArmorMechanicsAPI.setItem(equipment, slot, null);
            }
        }
    }

    public static void info(CommandSender sender) {
        PluginDescriptionFile desc =  ArmorMechanics.INSTANCE.getDescription();
        sender.sendMessage("" + GRAY + GOLD + BOLD + "Armor" + GRAY + BOLD + "Mechanics"
                + GRAY + ", v" + ITALIC + desc.getVersion());

        sender.sendMessage("  " + GRAY + SYM + GOLD + " Authors: " + GRAY + String.join(", ", desc.getAuthors()));
        sender.sendMessage("  " + GRAY + SYM + GOLD + " Command:" + GRAY + " /armormechanics");

        sender.sendMessage("  " + GRAY + SYM + GOLD + " Server: " + GRAY + Bukkit.getName() + " " + Bukkit.getVersion());
        sender.sendMessage("  " + GRAY + SYM + GOLD + " MechanicsCore: " + GRAY + MechanicsCore.getPlugin().getDescription().getVersion());
        sender.sendMessage("  " + GRAY + SYM + GOLD + " WeaponMechanics: " + GRAY + WeaponMechanics.getPlugin().getDescription().getVersion());
        sender.sendMessage("  " + GRAY + SYM + GOLD + " Java: " + GRAY + System.getProperty("java.version"));

        // Gets all supported plugins
        Set<String> softDepends = new LinkedHashSet<>(desc.getSoftDepend());
        softDepends.addAll(MechanicsCore.getPlugin().getDescription().getSoftDepend());
        softDepends.removeIf(name -> Bukkit.getPluginManager().getPlugin(name) == null);
        sender.sendMessage("  " + GRAY + SYM + GOLD + " Supported plugins: " + GRAY + String.join(", ", softDepends));
    }
}

File: src/main/java/com/cjcrafter/armormechanics/commands/Command.kt
@file:Suppress("UNCHECKED_CAST")

package com.cjcrafter.armormechanics.commands

import com.cjcrafter.armormechanics.ArmorMechanics
import com.cjcrafter.armormechanics.ArmorMechanicsAPI
import com.cjcrafter.armormechanics.events.ArmorGenerateEvent
import dev.jorel.commandapi.arguments.ArgumentSuggestions
import dev.jorel.commandapi.kotlindsl.anyExecutor
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.entitySelectorArgumentManyEntities
import dev.jorel.commandapi.kotlindsl.stringArgument
import dev.jorel.commandapi.kotlindsl.subcommand
import me.deecaad.core.commands.CommandHelpBuilder
import me.deecaad.core.commands.CustomMapArgument
import me.deecaad.core.file.simple.BooleanSerializer
import me.deecaad.core.file.simple.CsvSerializer
import me.deecaad.core.file.simple.StringSerializer
import me.deecaad.core.utils.AdventureUtil
import me.deecaad.core.utils.EnumUtil
import me.deecaad.core.utils.StringUtil
import me.deecaad.weaponmechanics.utils.CustomTag
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.Style
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Entity
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.inventory.EquipmentSlot

object Command {
    fun register() {
        val armorDataMapArgument =
            CustomMapArgument(
                "data",
                mapOf(
                    "dontEquip" to BooleanSerializer(),
                    "forceEquip" to BooleanSerializer(),
                    "preventRemove" to BooleanSerializer(),
                    "attachments" to CsvSerializer(StringSerializer()),
                )
            )

        commandAPICommand("am") {
            withAliases("armor", "armormechanics")
            withPermission("armormechanics.admin")
            withShortDescription("ArmorMechanics main command")

            subcommand("give") {
                withPermission("armormechanics.commands.give")
                withShortDescription("Gives the target(s) armor")

                entitySelectorArgumentManyEntities("targets")
                stringArgument("armor") {
                    replaceSuggestions(
                        ArgumentSuggestions.strings {
                            val options = ArmorMechanics.INSTANCE.armors.keys
                            options.toTypedArray()
                        }
                    )
                }
                withOptionalArguments(armorDataMapArgument)

                anyExecutor { sender, args ->
                    val targets = args["targets"] as List<Entity>
                    val armorTitle = args["armor"] as String
                    val data = args["data"] as? Map<String, Any> ?: mutableMapOf()

                    give(sender, targets, armorTitle, data)
                }
            }

            subcommand("get") {
                withPermission("armormechanics.commands.get")
                withShortDescription("Gives yourself armor")

                stringArgument("armor") {
                    replaceSuggestions(
                        ArgumentSuggestions.strings {
                            val options = ArmorMechanics.INSTANCE.armors.keys
                            options.toTypedArray()
                        }
                    )
                }
                withOptionalArguments(armorDataMapArgument)

                anyExecutor { sender, args ->
                    val armorTitle = args["armor"] as String
                    val data = args["data"] as? Map<String, Any> ?: mutableMapOf()

                    give(sender, listOf(sender as Player), armorTitle, data)
                }
            }

            subcommand("giveset") {
                withPermission("armormechanics.commands.giveset")
                withShortDescription("Gives the target(s) a set of armor")

                entitySelectorArgumentManyEntities("targets")
                stringArgument("set") {
                    replaceSuggestions(
                        ArgumentSuggestions.strings {
                            val options = ArmorMechanics.INSTANCE.sets.keys
                            options.toTypedArray()
                        }
                    )
                }
                withOptionalArguments(armorDataMapArgument)

                anyExecutor { sender, args ->
                    val targets = args["targets"] as List<Entity>
                    val setTitle = args["set"] as String
                    val data = args["data"] as? Map<String, Any> ?: mutableMapOf()

                    val set = ArmorMechanics.INSTANCE.sets[setTitle]
                    if (set == null) {
                        sender.sendMessage("${ChatColor.RED}Unknown set '$setTitle'")
                        return@anyExecutor
                    }

                    give(sender, targets, set.helmet ?: "", data)
                    give(sender, targets, set.chestplate ?: "", data)
                    give(sender, targets, set.leggings ?: "", data)
                    give(sender, targets, set.boots ?: "", data)
                }
            }

            subcommand("clear") {
                withPermission("armormechanics.commands.clear")
                withShortDescription("Clears the target(s) armor")

                entitySelectorArgumentManyEntities("targets")
                stringArgument("slots", optional = true) {
                    replaceSuggestions(
                        ArgumentSuggestions.strings {
                            val options = EnumUtil.getOptions(EquipmentSlot::class.java)
                            options.toTypedArray()
                        }
                    )
                }

                anyExecutor { sender, args ->
                    val targets = args["targets"] as List<Entity>
                    val slots = args["slots"] as? String

                    clear(sender, targets, slots)
                }
            }

            subcommand("reload") {
                withPermission("armormechanics.commands.reload")
                withShortDescription("Reloads ArmorMechanics configuration")

                anyExecutor { sender, args ->
                    ArmorMechanics.INSTANCE.reload().thenRun {
                        sender.sendMessage("${ChatColor.GREEN}Reloaded ArmorMechanics")
                    }
                }
            }

            val helpBuilder = CommandHelpBuilder(Style.style(NamedTextColor.GOLD), Style.style(NamedTextColor.GRAY))
            helpBuilder.register(this)
        }
    }

    fun give(sender: CommandSender, entities: List<Entity>, title: String, data: Map<String, Any>) {

        // Since we want to ignore spelling/capitalization errors, we should
        // make sure the given 'title' matches to an actual armor-title.
        var title = title
        val startsWith: MutableList<String> = ArrayList()
        val options: Set<String> = ArmorMechanics.INSTANCE.armors.keys
        for (temp in options) {
            if (temp.lowercase().startsWith(title.lowercase())) startsWith.add(title)
        }
        title = StringUtil.didYouMean(title, if (startsWith.isEmpty()) options else startsWith)
        val armor = ArmorMechanics.INSTANCE.armors[title]
        if (armor == null) {
            sender.sendMessage(ChatColor.RED.toString() + "Couldn't find armor '" + title + "'... Choose from " + options)
            return
        }

        val slot = ArmorMechanicsAPI.guessEquipmentSlot(armor)!!
        val dontEquip = data["dontEquip"] as? Boolean ?: false
        val force = data["forceEquip"] as? Boolean ?: false
        val preventRemove = data["preventRemove"] as? Boolean ?: false
        if (!force && preventRemove) {
            sender.sendMessage(ChatColor.RED.toString() + "When using preventRemove, forceEquip must also be enabled!")
        }
        for (entity in entities) {
            if (entity !is LivingEntity) continue
            val equipment = entity.equipment ?: continue

            // Let other plugins modify generated armor
            val clone = armor.clone()
            val event = ArmorGenerateEvent(sender, entity, clone, title, data)
            Bukkit.getPluginManager().callEvent(event)

            AdventureUtil.updatePlaceholders(entity as? Player, clone)

            if (!dontEquip && force) {
                if (preventRemove) {
                    CustomTag.PREVENT_REMOVE.setInteger(clone, 1)
                    equipment.setItem(slot, clone)
                    return
                }
                equipment.setItem(slot, clone)
                return
            }
            if (entity is Player) {
                val overflow = entity.inventory.addItem(clone)
                if (overflow.isNotEmpty()) {
                    sender.sendMessage(ChatColor.RED.toString() + entity.name + "'s inventory was full!")
                }
            }
        }
    }

    fun clear(sender: CommandSender, targets: List<Entity>, slot: EquipmentSlot?) {
        sender.sendMessage(ChatColor.GREEN.toString() + "Removing armor from " + targets.size + if (targets.size == 1) "entity" else "entities")
        for (target in targets) {
            if (!target.type.isAlive) continue
            val entity = target as LivingEntity
            val equipment = entity.equipment!!
            if (slot == null) {
                equipment.boots = null
                equipment.leggings = null
                equipment.chestplate = null
                equipment.helmet = null
            } else {
                equipment.setItem(slot, null)
            }
        }
    }
}
